### PR TITLE
session: run shutdown script, clean up activation env before exit

### PR DESCRIPTION
**Pull request by ahesford**
This came up on IRC and seems like it would be a useful addition.

### CODE REVIEW

**Comment by Consolatis** [comment]
Code LGTM, we should likely document it though. Also up to @johanmalm if we want to merge this before or after release.

**Comment by ahesford** [comment]
No rush to merge before release---I can go either way. I'll write up some documentation to go along with the change.

**Comment by johanmalm** [comment]
I like this in principle, but prefer to leave until after release.

`update_activation_env()` contains calls to both `dbus-update-activation-environment` and `systemctl` so it feels like we ought to call both on "unactivation" also.

https://github.com/labwc/labwc/blob/17bad48d87f5550dcdd16a03fd95b9bd9ebc73b3/src/config/session.c#L82

File: docs/labwc-config.5.scd [hunk 1]
```diff
@@ -48,6 +45,20 @@ sourced prior to running openbox.
...
+support) DISPLAY will be defined. This is the place for executing clients for
+handling background images, panels and other tasks that should run automatically
+when labwc launches.
+
+The *shutdown* file is executed as a shell script when labwc is preparing to
+terminate itself. All environment variables, including WAYLAND_DISPLAY and
+DISPLAY, will be available to the script. However, because the script runs
+asynchronously with other termination tasks, the shutdown file should not assume
+that the display will be usable. This is a perform any custom operations
+necessary to finalize a labwc session.
```
Comment (Consolatis) [inline docs/labwc-config.5.scd:60]: There seems to be missing something

**Comment by ahesford** [comment]
I split the dbus and systemd activation stuff into separate functions given the different set/unset behaviors for each. Suggestions welcome for improving this.

**Comment by Consolatis** [comment]
> I split the dbus and systemd activation stuff into separate functions given the different set/unset behaviors for each. Suggestions welcome for improving this.

Looks good to me.

I am not sure if `systemctl` might end up spawning a new dbus session as well if there was none. Before this PR it wasn't called if `DBUS_SESSION_BUS_ADDRESS` wasn't set while now it is. But again, not sure if that is an actual problem.

**Comment by ahesford** [comment]
Taking a different tack here:
- Restore the dependence of systemctl activation on a valid dbus session-bus address
- With https://github.com/labwc/labwc/pull/1546 pending, add some dynamic behavior that makes it easier to do the right thing if we add or remove variables in the future.

I probably don't need so many length checks, but we're concatenating strings so I took extra precautions... Comments welcome.

**Comment by johanmalm** [comment]
Could we simplify str_concat by using the common/buf.h helpers?

https://github.com/labwc/labwc/blob/1aa906e0de740a6af96c7365e03c137f84bbc414/include/common/buf.h#L46

Or even add a create a buf_str_join(buf, string-array, delim) where the delim could be “ “ and “= “ for the two cases.

**Comment by ahesford** [comment]
I generalized the join and moved it alongside the other string helpers. Trying to use the buffer struct for this might be a little awkward, and seems like it would involve several reallocs.

**Comment by ahesford** [comment]
Now that 0.7.1 is out, I think this is ready for review.

**Review by johanmalm — COMMENTED** [review]
Two minor comments. Other than those LGTM.
Nice improvement.
Great effort with `str_join()`

**Comment by ahesford** [comment]
Changes adopted

**Pull request merged**